internal/graph/rate_limited_check_resolver.go

```diff
@@ -0,0 +1,176 @@
+package graph
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/pkg/telemetry"
+)
+
+// RateLimitedCheckResolverConfig encapsulates configuration for rate limited check resolver
+type RateLimitedCheckResolverConfig struct {
+	TimerTickerFrequency time.Duration
+	MediumPriorityLevel  uint32
+	MediumPriorityShaper uint32
+	LowPriorityLevel     uint32
+	LowPriorityShaper    uint32
+}
+
+// RateLimitedCheckResolver will prioritize requests with fewer dispatches over
+// requests with more dispatches.
+// Initially, request's dispatches will not be throttled and will be processed
+// immediately. As the number of dispatches increases (to beyond the config's MediumPriorityLevel),
+// selectively number (configure via config's MediumPriorityShaper) will be throttled by placing the
+// dispatch in the medium priority queue. One item from the medium priority queue will be processed
+// every ticker across the entire server. Request's dispatches will gradually be more frequently throttled as
+// the number of dispatches increase until every dispatch will be throttled by being placed in the medium priority
+// queue. When the number of request dispatches is above the LowPriorityLevel, the dispatches are placed
+// in the low priority queue. One item form the low priority queue will be processed every Nth ticker
+// (configured via config's LowPriorityShaper parameter).
+// This allows a check / list objects request to be gradually throttled.
+type RateLimitedCheckResolver struct {
+	delegate         CheckResolver
+	config           RateLimitedCheckResolverConfig
+	ticker           *time.Ticker
+	lowPriorityQueue chan bool
+	medPriorityQueue chan bool
+	done             chan bool
+
+	// these are helper config value to reduce calculation needed for medium shaper
```

Comment:
nit: this comment isn't adding any value to me, i suggest removing it and adding godocs in lines 17 and 18 because i still don't know what "shaper" means

Reply:
Rename to rate